plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1238 @@
+
+	fu_progress_step_done(progress);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_probe(FuDevice *device, GError **error)
+{
+#if G_USB_CHECK_VERSION(0, 3, 3)
+	FuSteelseriesSonic *self = FU_STEELSERIES_SONIC(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	GUsbInterface *iface = NULL;
+	GUsbEndpoint *ep = NULL;
+	guint8 iface_id;
+	guint8 ep_id;
+	guint16 packet_size;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+
+	/* FuUsbDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_steelseries_sonic_parent_class)->probe(device, error))
+		return FALSE;
+
+	ifaces = g_usb_device_get_interfaces(usb_device, error);
+	if (ifaces == NULL || ifaces->len == 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "update interface not found");
+		return FALSE;
+	}
+
+	/* use the last interface for interrupt transfer */
+	iface_id = ifaces->len - 1;
+
+	iface = g_ptr_array_index(ifaces, iface_id);
+
+	endpoints = g_usb_interface_get_endpoints(iface);
+	/* expecting to have only one endpoint for communication */
+	if (endpoints == NULL || endpoints->len != 1) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "endpoint not found");
+		return FALSE;
+	}
+
+	ep = g_ptr_array_index(endpoints, 0);
+	ep_id = g_usb_endpoint_get_address(ep);
+	packet_size = g_usb_endpoint_get_maximum_packet_size(ep);
+
+	self->iface_idx = iface_id;
+	self->ep = ep_id;
+	self->in_size = packet_size;
+
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), iface_id);
+
+	/* success */
+	return TRUE;
+#else
+	/* failed */
+	g_set_error_literal(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "this version of GUsb is not supported");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_steelseries_sonic_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	SteelseriesSonicChip chip;
+	g_autoptr(FwupdRequest) request = NULL;
+	g_autofree gchar *msg = NULL;
+
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 50);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 50);
+
+	/* mouse */
+	chip = STEELSERIES_SONIC_CHIP_MOUSE;
+	if (!fu_steelseries_sonic_restart(device, chip, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "failed to restart chip %u: ", chip);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* dongle (nordic, holtek; same command) */
+	chip = STEELSERIES_SONIC_CHIP_HOLTEK;
+	if (!fu_steelseries_sonic_restart(device, chip, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "failed to restart chip %u: ", chip);
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* the user has to do something */
+	msg = g_strdup_printf("%s needs to be manually restarted to complete the update. "
+			      "Please unplug the dongle and then re-plug it.",
```

Comment:
i think dongle is one of those words we're not supposed to use any more. perhaps "receiver" instead?

Reply:
Arf, do we? I was not aware about that word.

Reply:
https://www.urbandictionary.com/define.php?term=dongle